src/repr/src/row.proto

```diff
@@ -0,0 +1,144 @@
+        ProtoDate date = 16;
+        ProtoTime time = 17;
+        ProtoTimestamp timestamp = 18;
+        ProtoInterval interval = 19;
+        ProtoArray array = 20;
+        ProtoRow list = 21;
+        ProtoDict dict = 22;
+        ProtoNumeric numeric = 23;
+        bytes uuid = 24;
+    }
+}
+
+enum ProtoDatumOther {
+    // It's generally good practice to make id 0 (the default if the field is
+    // unset) in proto enums be an Unknown sentinel. This allows for
+    // distinguishing between unset and any of the enum variants.
+    //
+    // This enum is initially used only in a oneof, which means we can
+    // distinguish unset without this sentinel. But stick one in here anyway,
+    // in case this enum gets used somewhere else in the future.
+    Unknown = 0;
+    JsonNull = 1;
+    Dummy = 2;
+}
+
+message ProtoDate {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+}
+
+message ProtoTime {
+    // The number of seconds since midnight
+    uint32 secs = 1;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 2;
+}
+
+message ProtoTimestamp {
+    // Year
+    int32 year = 1;
+    // Day-of-year (0..365)
+    uint32 ordinal = 2;
+    // The number of seconds since midnight
+    uint32 secs = 3;
+    // Additional fractional seconds since midnight in nanosecond granularity.
+    // This can can exceed 1,000,000,000 in order to represent the leap second.
+    uint32 nanos = 4;
+    // If true, this timestamp is in UTC. If false, this timestamp is zoneless.
+    bool is_tz = 5;
+}
+
+message ProtoInterval {
+    // A possibly negative number of months.
+    int32 months = 1;
+    // A 128-bit nanosecond count, represented as a 64-bit lo and hi.
+    //
+    // TODO: Why do our in-mem intervals use this representation? I've seen
+    // month/day/nanos in the past becuase a day is not always the same number
+    // of nanos.
+    int64 duration_lo = 2;
+    int64 duration_hi = 3;
+}
+
+message ProtoArray {
+    // All array elements flattened into 1 dimension, encoded in row-major
+    // order.
+    ProtoRow elements = 1;
+    // A list of metadata for each dimension in the array. Each dimension has a
+    // lower bound (the index at which the dimension begins) and the length of
+    // the dimension (the number of elements in that dimension). For a 3x4
+    // matrix, for example, you'd have two entries in the dims array, the first
+    // with length 3 and the second with length 4. ATM the lower bound for each
+    // dimension is always 1, but Postgres technically lets you choose any lower
+    // bound you like for each dimension.
+    repeated ProtoArrayDimension dims = 2;
+}
+
+message ProtoArrayDimension {
+    uint64 lower_bound = 1;
+    uint64 length = 2;
+}
+
+message ProtoDict {
+    repeated ProtoDictElement elements = 1;
```

Comment:
is there a reason to use `repeated` over native maps here? https://developers.google.com/protocol-buffers/docs/proto3#maps